### PR TITLE
Simplify constructors/operators of AffineConstraints::ConstraintLine.

### PR DESCRIPTION
**Pull request by bangerth**
In `AffineConstraints::close()`, we sort objects of type `AffineConstraints::ConstraintLine`. For this, it is useful if objects are move constructible and move assignable. This patch makes sure this is so by adding move constructors and move operators.

Interestingly, the copy constructor and copy operator of this class are templatized to take objects that for all practical purposes need to look exactly like `ConstraintLine`, down to the level of naming of their member variables. That's definitely weird -- I have no idea why this is the case, so I removed it and now only allow `ConstraintLine` objects as arguments. This seems to work when compiling the library; let's see whether that also works for the test suite and if someone remembers why we did it that way.

### CODE REVIEW

**Comment by drwells** [comment]
It looks like we added these templates in #11678: I think they're necessary for mixing scalar types.

**Comment by bangerth** [comment]
I fixed the test as well (using modifications in the `copy_from()` function -- so there is not external breakage).

**Comment by bangerth** [comment]
Part of #15375.

**Review by drwells — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/lac/affine_constraints.h [hunk 1]
```diff
@@ -1658,15 +1658,24 @@ class AffineConstraints : public Subscriptor
...
      * Copy constructor.
      */
-    template <typename ConstraintLineType>
-    ConstraintLine(const ConstraintLineType &other);
+    ConstraintLine(const ConstraintLine &other) = default;
+
+    /**
+     * Move constructor.
+     */
+    ConstraintLine(ConstraintLine &&other) = default;
```
Comment (drwells) [inline include/deal.II/lac/affine_constraints.h:1666]: We should mark this as `noexcept` since it cannot throw an exception.

Comment (bangerth) [reply include/deal.II/lac/affine_constraints.h]: So done!

**Review by kronbichler — APPROVED** [review]

**Comment by masterleinad** [comment]
```
/home/runner/work/dealii/dealii/include/deal.II/lac/affine_constraints.h:2531: warning: no matching class member found for 
  template < number >
  AffineConstraints< number >::ConstraintLine::ConstraintLine(const size_type &index, const typename AffineConstraints< number >::ConstraintLine::Entries &entries, const number &inhomogeneity)
Possible candidates:
  'AffineConstraints< number >::ConstraintLine::ConstraintLine(const size_type &index=numbers::invalid_dof_index, const Entries &entries={}, const number &inhomogeneity=0.0)'
  'AffineConstraints< number >::ConstraintLine::ConstraintLine(const ConstraintLine &other)=default'
  'AffineConstraints< number >::ConstraintLine::ConstraintLine(ConstraintLine &&other) noexcept=default'
```

**Comment by bangerth** [comment]
That's a bug in doxygen :-( Let's see whether this variation works. I've taken the liberty to also make a scalar argument by-value instead of by-reference.

**Pull request merged**